Flaky test: AllMetricsOneFunction

```
--- FAIL: Test (0.15s)
    --- FAIL: Test/AllMetricsOneFunction (0.07s)
        tlogger.go:116: INFO server.go:629 [core] [Server #19] Server created  (t=+1.130319ms)
        tlogger.go:116: INFO clientconn.go:177 [core] [Channel #20] Channel created  (t=+1.558427ms)
        tlogger.go:116: INFO clientconn.go:1600 [core] [Channel #20] original dial target is: "whatever:///127.0.0.1:45395"  (t=+1.668828ms)
        tlogger.go:116: INFO clientconn.go:1607 [core] [Channel #20] parsed dial target is: {Scheme:whatever Authority: URL:{Scheme:whatever Opaque: User: Host: Path:/127.0.0.1:45395 RawPath: OmitHost:false ForceQuery:false RawQuery: Fragment: RawFragment:}}  (t=+1.824031ms)
        tlogger.go:116: INFO clientconn.go:1707 [core] [Channel #20] Channel authority set to "127.0.0.1:45395"  (t=+1.928833ms)
        tlogger.go:116: INFO resolver_conn_wrapper.go:196 [core] [Channel #20] Resolver state updated: {
              "Addresses": [
                {
                  "Addr": "127.0.0.1:45395",
                  "ServerName": "",
                  "Attributes": null,
                  "BalancerAttributes": null,
                  "Type": 0,
                  "Metadata": null
                }
              ],
              "ServiceConfig": null,
              "Attributes": null
            } (resolver returned new addresses)  (t=+2.188437ms)
        tlogger.go:116: INFO balancer_conn_wrappers.go:276 [core] [Channel #20] Channel switches to new LB policy "pick_first"  (t=+2.37184ms)
        tlogger.go:116: INFO clientconn.go:759 [core] [Channel #20 SubChannel #21] Subchannel created  (t=+2.5[336](https://github.com/grpc/grpc-go/actions/runs/4824796293/jobs/8595006405?pr=6228#step:8:337)43ms)
        tlogger.go:116: INFO clientconn.go:451 [core] [Channel #20] Channel Connectivity change to CONNECTING  (t=+2.629444ms)
        tlogger.go:116: INFO server.go:817 [core] [Server #19 ListenSocket #22] ListenSocket created  (t=+2.788947ms)
        tlogger.go:116: INFO clientconn.go:1141 [core] [Channel #20 SubChannel #21] Subchannel Connectivity change to CONNECTING  (t=+3.064652ms)
        tlogger.go:116: INFO clientconn.go:1255 [core] [Channel #20 SubChannel #21] Subchannel picks a new address "127.0.0.1:45395" to connect  (t=+3.254755ms)
        tlogger.go:116: INFO clientconn.go:1141 [core] [Channel #20 SubChannel #21] Subchannel Connectivity change to READY  (t=+4.581777ms)
        tlogger.go:116: INFO clientconn.go:451 [core] [Channel #20] Channel Connectivity change to READY  (t=+4.880182ms)
        e2e_test.go:1019: got unexpected viewInformation for metric grpc.io/server/completed_rpcs, diff (-got, +want):   (*opencensus.viewInformation)(
            - 	&{
            - 		aggType:    s"Count",
            - 		aggBuckets: []float64{},
            - 		desc:       "Number of completed RPCs by method and status.",
            - 		tagKeys:    []tag.Key{{name: "grpc_server_method"}, {name: "grpc_server_status"}},
            - 		rows:       []*view.Row{s"{ { {grpc_server_method grpc.testing.TestService/FullDuplexCall}"...},
            - 	},
            + 	&{
            + 		aggType:    s"Count",
            + 		aggBuckets: []float64{},
            + 		desc:       "Number of completed RPCs by method and status.",
            + 		tagKeys:    []tag.Key{{name: "grpc_server_method"}, {name: "grpc_server_status"}},
            + 		rows: []*view.Row{
            + 			s"{ { {grpc_server_method grpc.testing.TestService/UnaryCall}{grpc"...,
            + 			s"{ { {grpc_server_method grpc.testing.TestService/FullDuplexCall}"...,
            + 		},
            + 	},
              )
        tlogger.go:116: INFO clientconn.go:451 [core] [Channel #20] Channel Connectivity change to SHUTDOWN  (t=+55.584632ms)
        tlogger.go:116: INFO clientconn.go:1141 [core] [Channel #20 SubChannel #21] Subchannel Connectivity change to SHUTDOWN  (t=+55.955638ms)
        tlogger.go:116: INFO clientconn.go:1468 [core] [Channel #20 SubChannel #21] Subchannel deleted  (t=+56.110[341](https://github.com/grpc/grpc-go/actions/runs/4824796293/jobs/8595006405?pr=6228#step:8:342)ms)
        tlogger.go:116: INFO clientconn.go:1095 [core] [Channel #20] Channel deleted  (t=+56.201042ms)
        tlogger.go:116: INFO server.go:768 [core] [Server #19 ListenSocket #22] ListenSocket deleted  (t=+57.087057ms)
FAIL
FAIL	google.golang.org/grpc/stats/opencensus	0.831s
```	

https://github.com/grpc/grpc-go/actions/runs/4824796293/jobs/8595006405?pr=6228

--- a/stats/opencensus/e2e_test.go
+++ b/stats/opencensus/e2e_test.go
@@ -238,16 +238,21 @@ func distributionDataLatencyCount(vi *viewInformation, countWant int64, wantTags
 }
 
 // waitForServerCompletedRPCs waits until both Unary and Streaming metric rows
-// appear for server completed RPC's view (by checking for length of rows to be
-// 2). Returns an error if both the Unary and Streaming metric not found within
-// the passed context's timeout.
-func waitForServerCompletedRPCs(ctx context.Context, fe *fakeExporter) error {
+// appear for server completed RPC's view. Returns an error if the Unary and
+// Streaming metric are not found within the passed context's timeout.
+func waitForServerCompletedRPCs(ctx context.Context) error {
 	for ; ctx.Err() == nil; <-time.After(time.Millisecond) {
 		rows, err := view.RetrieveData("grpc.io/server/completed_rpcs")
 		if err != nil {
 			continue
 		}
-		if len(rows) == 2 {
+		m := make(map[string]bool)
+		for _, row := range rows {
+			for _, tag := range row.Tags {
+				m[tag.Value] = true
+			}
+		}
+		if m["grpc.testing.TestService/UnaryCall"] && m["grpc.testing.TestService/FullDuplexCall"] {
 			return nil
 		}
 	}
@@ -1008,7 +1013,7 @@ func (s) TestAllMetricsOneFunction(t *testing.T) {
 	// Streaming calls with respect to the RPC returning client side. Thus, add
 	// a sync point at the global view package level for these two rows to be
 	// recorded, which will be synchronously uploaded to exporters right after.
-	if err := waitForServerCompletedRPCs(ctx, fe); err != nil {
+	if err := waitForServerCompletedRPCs(ctx); err != nil {
 		t.Fatal(err)
 	}
 	view.Unregister(allViews...)

--- a/stats/opencensus/e2e_test.go
+++ b/stats/opencensus/e2e_test.go
@@ -237,6 +237,23 @@ func distributionDataLatencyCount(vi *viewInformation, countWant int64, wantTags
 	return nil
 }
 
+// waitForServerCompletedRPCs waits until both Unary and Streaming metric rows
+// appear for server completed RPC's view (by checking for length of rows to be
+// 2). Returns an error if both the Unary and Streaming metric not found within
+// the passed context's timeout.
+func waitForServerCompletedRPCs(ctx context.Context, fe *fakeExporter) error {
+	for ; ctx.Err() == nil; <-time.After(time.Millisecond) {
+		rows, err := view.RetrieveData("grpc.io/server/completed_rpcs")
+		if err != nil {
+			continue
+		}
+		if len(rows) == 2 {
+			return nil
+		}
+	}
+	return fmt.Errorf("timeout when waiting for Unary and Streaming rows to be present for \"grpc.io/server/completed_rpcs\"")
+}
+
 // TestAllMetricsOneFunction tests emitted metrics from gRPC. It registers all
 // the metrics provided by this package. It then configures a system with a gRPC
 // Client and gRPC server with the OpenCensus Dial and Server Option configured,
@@ -987,10 +1004,13 @@ func (s) TestAllMetricsOneFunction(t *testing.T) {
 			},
 		},
 	}
-	// Unregister all the views. Unregistering a view causes a synchronous
-	// upload of any collected data for the view to any registered exporters.
-	// Thus, after this unregister call, the exporter has the data to make
-	// assertions on immediately.
+	// Server Side stats.End call happens asynchronously for both Unary and
+	// Streaming calls with respect to the RPC returning client side. Thus, add
+	// a sync point at the global view package level for these two rows to be
+	// recorded, which will be synchronously uploaded to exporters right after.
+	if err := waitForServerCompletedRPCs(ctx, fe); err != nil {
+		t.Fatal(err)
+	}
 	view.Unregister(allViews...)
 	// Assert the expected emissions for each metric match the expected
 	// emissions.

--- a/stats/opencensus/e2e_test.go
+++ b/stats/opencensus/e2e_test.go
@@ -238,7 +238,7 @@ func distributionDataLatencyCount(vi *viewInformation, countWant int64, wantTags
 }
 
 // waitForServerCompletedRPCs waits until both Unary and Streaming metric rows
-// appear, in two seperate rows, for server completed RPC's view. Returns an
+// appear, in two separate rows, for server completed RPC's view. Returns an
 // error if the Unary and Streaming metric are not found within the passed
 // context's timeout.
 func waitForServerCompletedRPCs(ctx context.Context) error {

--- a/stats/opencensus/e2e_test.go
+++ b/stats/opencensus/e2e_test.go
@@ -238,22 +238,30 @@ func distributionDataLatencyCount(vi *viewInformation, countWant int64, wantTags
 }
 
 // waitForServerCompletedRPCs waits until both Unary and Streaming metric rows
-// appear for server completed RPC's view. Returns an error if the Unary and
-// Streaming metric are not found within the passed context's timeout.
+// appear, in two seperate rows, for server completed RPC's view. Returns an
+// error if the Unary and Streaming metric are not found within the passed
+// context's timeout.
 func waitForServerCompletedRPCs(ctx context.Context) error {
 	for ; ctx.Err() == nil; <-time.After(time.Millisecond) {
 		rows, err := view.RetrieveData("grpc.io/server/completed_rpcs")
 		if err != nil {
 			continue
 		}
-		m := make(map[string]bool)
+		unaryFound := false
+		streamingFound := false
 		for _, row := range rows {
 			for _, tag := range row.Tags {
-				m[tag.Value] = true
+				if tag.Value == "grpc.testing.TestService/UnaryCall" {
+					unaryFound = true
+					break
+				} else if tag.Value == "grpc.testing.TestService/FullDuplexCall" {
+					streamingFound = true
+					break
+				}
 			}
-		}
-		if m["grpc.testing.TestService/UnaryCall"] && m["grpc.testing.TestService/FullDuplexCall"] {
-			return nil
+			if unaryFound && streamingFound {
+				return nil
+			}
 		}
 	}
 	return fmt.Errorf("timeout when waiting for Unary and Streaming rows to be present for \"grpc.io/server/completed_rpcs\"")

